utils/bfd.ml

```diff
@@ -0,0 +1,540 @@
+    seek (Int64.add e_lfanew (Int64.of_int header_size));
+    let buf          = read_bytes size_of_optional_header in
+    let base_of_code = get_int32 buf 20 in
+    let image_base   = get_word  buf (match bitness with X64 -> 24 | X86 -> 28) in
+    {base_of_code; image_base}
+
+  type section =
+    {
+      name: string;
+      virtual_size: int;
+      virtual_address: int32;
+      size_of_raw_data: int;
+      pointer_to_raw_data: int;
+    }
+
+  let section_header_size = 40
+
+  let sections =
+    seek Int64.(add e_lfanew (of_int (header_size + size_of_optional_header)));
+    let mk buf =
+      let name                = name_at ~max:8 buf 0 in
+      let virtual_size        = get_int   buf 8 in
+      let virtual_address     = get_int32 buf 12 in
+      let size_of_raw_data    = get_int   buf 16 in
+      let pointer_to_raw_data = get_int   buf 20 in
+      {name; virtual_size; virtual_address; size_of_raw_data; pointer_to_raw_data}
+    in
+    Array.init number_of_sections (fun _ -> mk (read_bytes section_header_size))
+
+  type symbol =
+    {
+      name: string;
+      address: int64;
+    }
+
+  let read_section_body {size_of_raw_data; pointer_to_raw_data; _} =
+    seek (Int64.of_int pointer_to_raw_data);
+    read_bytes size_of_raw_data
+
+  let find_section sectname =
+    array_find (function ({name; _} : section) -> name = sectname) sections
+
+  let symbols =
+    match find_section ".exptbl" with
+    | None -> [| |]
+    | Some ({virtual_address; _} as exptbl) ->
+        let data       = read_section_body exptbl in
+        let numexports = Int64.to_int (get_word data 0) in
+        Array.init numexports (fun i ->
+            let address = get_word data (word_size + word_size * 2 * i) in
+            let nameoff = get_word data (word_size + word_size * 2 * i + word_size) in
+            let name    =
+              let off =
+                Int64.(sub nameoff (add (unsigned_of_int32 virtual_address) image_base))
+              in
+              name_at data (Int64.to_int off)
+            in
+            {name; address}
+          )
+
+  let symbol_offset symname =
+    match array_find (function {name; _} -> name = symname) symbols with
+    | None -> None
+    | Some {address; _} ->
+        Some Int64.(sub address (add (unsigned_of_int32 base_of_code) image_base))
+
+  let defines_symbol symname =
+    Array.exists (fun {name; _} -> name = symname) symbols
+end
+
+module type T = functor (_ : READ) -> S
+
+let guess ic =
+  seek_in ic 0;
+  let magic = really_input_string ic 4 in
+  match magic.[0], magic.[1], magic.[2], magic.[3] with
+  | '\x7F', 'E', 'L', 'F' ->
+      (module ELF : T)
+  | '\xFE', '\xED', '\xFA', '\xCE'
+  | '\xCE', '\xFA', '\xED', '\xFE'
+  | '\xFE', '\xED', '\xFA', '\xCF'
+  | '\xCF', '\xFA', '\xED', '\xFE' ->
+      (module Mach_O : T)
+  | 'M', 'Z', _, _ ->
+      (module FlexDLL : T)
+  | _ ->
+      raise (Error (Unrecognized magic))
+
+type t =
+  (module S)
```

Comment:
I agree the design is a bit unusual. Using IO effects to define toplevel module items does not shock me personally, especially because I think of the functor as a function that takes the input channel as argument, and the toplevel items as "let-in"'s inside this function.

Originally I had a "decoder" object threaded pretty much through everywhere, but I felt it added quite a bit of noise to the code. I find the current version easy to follow and easy to reason about, if somewhat unusual in style.

But if it makes the code harder to understand for other people, it makes sense to switch to a more "conventional" style. I want to think about it a bit before making any changes.

Reply:
OK, after thinking about it a bit, I rewrote it in a style along the suggested lines; let me know what you think! Cheers.

Reply:
I agree that the new style is easier to follow, thanks! In terms of simplicity, I was a bit puzzled by `mutable buf : bytes; (* buf is never mutated *)` (if it is never mutated, then why is it mutable?), and I have the impression that it would be simpler to reason about a version where `load_bytes` does not mutate the decoder but instead just return the bytes, at the cost of writing `get_uint16 d b offset` instead of `get_uint16 d offset` as today. Another thing that would be simplified is the idiom `(load_bytes d (unsigned_of_int32 stroff) strsize; d.buf)`, where the projection would not be necessary anymore.

Reply:
Yes, apologies for the confusing comment. What it means is that the `bytes` value stored in the `buf` field is never mutated, the only reason it is not a string is because we only have the `get_int*` functions in `Bytes`.

I hear what you say about separating the buffer from the decoder. It will mean one more argument to pass around, but maybe it will end up being simpler like that. I will give it a shot.

Reply:
Pushed a commit with the suggested change. Indeed, it is a nice improvement, I'm quite happy with the result.